contracts/contracts/strategies/sonic/SonicSwapXAMOStrategy.sol

```diff
@@ -0,0 +1,738 @@
+     * @param _wsAmount Amount of Wrapped S (wS) to swap into the pool.
+     */
+    function swapAssetsToPool(uint256 _wsAmount)
+        external
+        onlyStrategist
+        nonReentrant
+        improvePoolBalance
+        skimPool
+    {
+        require(_wsAmount > 0, "Must swap something");
+
+        // 1. Partially remove liquidity so there’s enough wS for the swap
+
+        // Calculate how much pool LP tokens to burn to get the required amount of wS tokens back
+        uint256 lpTokens = _calcTokensToBurn(_wsAmount);
+        require(lpTokens > 0, "No LP tokens to burn");
+
+        _withdrawFromGaugeAndPool(lpTokens);
+
+        // 2. Swap wS for OS against the pool
+        // Swap exact amount of wS for OS against the pool
+        // There can be a dust amount of wS left in the strategy as the burn of the pool's LP tokens is rounded up
+        _swapExactTokensForTokens(_wsAmount, ws, os);
+
+        // 3. Burn all the OS left in the strategy from the remove liquidity and swap
+        uint256 osToBurn = IERC20(os).balanceOf(address(this));
+        IVault(vaultAddress).burnForStrategy(osToBurn);
+
+        // Ensure solvency of the vault
+        _solvencyAssert();
+
+        // Emit event for the burnt OS tokens
+        emit Withdrawal(os, pool, osToBurn);
+        // Emit event for the swap
+        emit SwapAssetsToPool(_wsAmount, lpTokens, osToBurn);
+    }
+
+    /**
+     * @notice Used when there is more wS than OS in the pool.
+     * OS is minted and swapped for wS against the pool,
+     * more OS is minted and added back into the pool with the swapped out wS.
+     * The OS/wS price is > 1.0 so OS is being sold at a premium.
+     * @param _osAmount Amount of OS to swap into the pool.
+     */
+    function swapOTokensToPool(uint256 _osAmount)
+        external
+        onlyStrategist
+        nonReentrant
+        improvePoolBalance
+        skimPool
+    {
+        require(_osAmount > 0, "Must swap something");
+
+        // 1. Mint OS so it can be swapped into the pool
+
+        // There shouldn't be any OS in the strategy but just in case
+        uint256 osInStrategy = IERC20(os).balanceOf(address(this));
+        require(_osAmount >= osInStrategy, "Too much OS in strategy");
+        uint256 osToMint = _osAmount - osInStrategy;
+
+        // Mint the required OS tokens to this strategy
+        IVault(vaultAddress).mintForStrategy(osToMint);
+
+        // 2. Swap OS for wS against the pool
+        _swapExactTokensForTokens(_osAmount, os, ws);
+
+        // The wS is from the swap and any wS that was sitting in the strategy
+        uint256 wsDepositAmount = IERC20(ws).balanceOf(address(this));
+
+        // 3. Add wS and OS back to the pool in proportion to the pool's reserves
+        (uint256 osDepositAmount, uint256 lpTokens) = _deposit(wsDepositAmount);
+
+        // Emit event for the minted OS tokens
+        emit Deposit(os, pool, osToMint + osDepositAmount);
```

Comment:
I am not sure adding these 2 amounts together are correct. The `osToMint` was part of the swap transaction and there to swap OS for wS. Once we have the wS the corresponding OS part is minted and deposited. 

Hmm I think I see the thought process, so instead of emitting 2 deposit events 1 for wS amount and 1 for OS amount, we emit only 1 with OS amounts combined.

Reply:
I guess this way of doing it is better, since even though we deposit ws&OS into the pool, no WS assets have really moved from the Vault. Only OS has been minted, swapped and deposited. I think the way you did it is more correct.

Maybe a comment to explain the thought process?

Reply:
The thought process was to not emit three Deposit events. Instead, just emit one Deposit event for wS and the other for OS. But as we mint OS twice, the amounts are added together